### PR TITLE
Log type that has failed to deserialize itself from a request

### PR DESCRIPTION
**Pull request by Kobzol**
Should help with debugging deserialization errors.

### CODE REVIEW

**Comment by Kobzol** [comment]
I pushed a different version that also logs the response body (otherwise it's hard to detect what is causing the error), at the cost of deserializing it to a string before parsing it as a JSON. Let me know what you think (the previous simpler commit is [here](https://github.com/rust-lang/sync-team/commit/77542c0e3cc1ef1e0683cd374a2c7e0340f49e75)).

File: src/utils.rs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        }
+    }
+
+    fn json_annotated<T: DeserializeOwned>(self) -> anyhow::Result<T> {
+        let text = self.text()?;
+        serde_json::from_str::<T>(&text).with_context(|| {
+            format!(
+                "Cannot deserialize type `{}` from\n{}",
+                std::any::type_name::<T>(),
+                text
```
Comment (rylev) [inline src/utils.rs:27]: Maybe we could try to deserialize to `serde_json::Value` first so that if it's valid JSON we'll at least know and be able to format a bit better than plain text.

File: src/utils.rs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+            Ok(_) => Ok(self),
+            Err(err) => {
+                let body = self.text()?;
+                Err(err).context(format!("Body: {:?}", body))
+            }
+        }
+    }
+
+    fn json_annotated<T: DeserializeOwned>(self) -> anyhow::Result<T> {
+        let text = self.text()?;
```
Comment (rylev) [inline src/utils.rs:22]: We should probably add some context here that this is the response body text - I'm not sure the error here will be particularly clear otherwise.

File: src/utils.rs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    fn custom_error_for_status(self) -> anyhow::Result<Response>;
+    fn json_annotated<T: DeserializeOwned>(self) -> anyhow::Result<T>;
+}
+
+impl ResponseExt for Response {
+    fn custom_error_for_status(self) -> anyhow::Result<Response> {
+        match self.error_for_status_ref() {
+            Ok(_) => Ok(self),
+            Err(err) => {
+                let body = self.text()?;
```
Comment (rylev) [inline src/utils.rs:15]: Same as below.

File: src/utils.rs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+                Err(err).context(format!("Body: {:?}", body))
```
Comment (rylev) [inline src/utils.rs:16]: This would put the "Body" error as the top error type. Maybe the following:

```suggestion
                anyhow!("Body for error: {body:?}").context(err)
```

File: src/utils.rs (hunk 1 above)
Comment (Kobzol) [reply src/utils.rs:27]: The error message isn't really much better if we parse from `serde_json::Value`, since `serde` by default doesn't include the "path" to the error in the error message.

File: src/utils.rs (hunk 4 above)
Comment (Kobzol) [reply src/utils.rs:16]: Btw, the wording is weird if I switch the contexts:
```
[2023-12-08T17:22:00Z ERROR sync_team] failed: HTTP status client error (400 Bad Request) for url (https://httpstat.us/400)
[2023-12-08T17:22:00Z ERROR sync_team] caused by: Body for error: "400 Bad Request"
```
it's not "caused by" the custom body content.

File: src/utils.rs (hunk 1 above)
Comment (rylev) [reply src/utils.rs:27]: Yea but the formatting of `serde_json::Value` is nicer than a plain string which can be hard to parse by eye.

**Review by rylev — APPROVED** [review]
Approving this since the feedback here shouldn't be considered blocking.

File: src/utils.rs (hunk 1 above)
Comment (Kobzol) [reply src/utils.rs:27]: That's true, although when you get a message like "missing field at column 1000", after formatting to `serde_json::Value`, there won't be any column 1000 in the debug output anymore :) But it's probably more useful anyway to format the response as JSON than to hunt column numbers on a very long line of text.

**Comment by Kobzol** [comment]
Oh, sorry, I actually forgot to push the JSON formatting change, I think that it's a good idea.
